share-coordinator/src/main/java/org/apache/kafka/coordinator/share/PersisterStateBatchCombiner.java

```diff
@@ -0,0 +1,394 @@
+
+    public PersisterStateBatchCombiner(
+        List<PersisterStateBatch> batchesSoFar,
+        List<PersisterStateBatch> newBatches,
+        long startOffset
+    ) {
+        initializeCombinedList(batchesSoFar, newBatches);
+        finalBatchList = new ArrayList<>(combinedBatchList.size() * 2);   // heuristic size
+        this.startOffset = startOffset;
+    }
+
+    private void initializeCombinedList(List<PersisterStateBatch> batchesSoFar, List<PersisterStateBatch> newBatches) {
+        boolean soFarEmpty = batchesSoFar == null || batchesSoFar.isEmpty();
+        boolean newBatchesEmpty = newBatches == null || newBatches.isEmpty();
+
+        if (soFarEmpty && newBatchesEmpty) {
+            combinedBatchList = new ArrayList<>();
+        } else if (soFarEmpty) {
+            combinedBatchList = newBatches;
+        } else if (newBatchesEmpty) {
+            combinedBatchList = batchesSoFar;
+        } else {
+            combinedBatchList = new ArrayList<>(batchesSoFar.size() + newBatches.size());
+            combinedBatchList.addAll(batchesSoFar);
+            combinedBatchList.addAll(newBatches);
+        }
+    }
+
+    /**
+     * Algorithm: Merge current state batches and new batches into a single non-overlapping batch list.
+     * Input: batchesSoFar, newBatches, startOffset
+     * Output: combined list with non-overlapping batches (finalBatchList)
+     * <p>
+     * - Add both currentBatches and newBatches into a single list combinedBatchList
+     * - Remove/prune any batches from the combinedBatchList:
+     * -    if batch.lastOffset < startOffset then remove batch from combinedBatchList
+     * -    else if batch.firstOffset > startOffset then we will keep the batch
+     * -    else if batch.firstOffset <= startOffset <= batch.lastOffset then keep [startOffset, batch.lastOffset] part only and discard rest.
+     * - create a treeset sortedBatches using pruned combinedBatchList
+     * - do repeat until all batches in sortedSet are non-overlapping:
+     * -    find first 2 overlapping batches in sortedBatches set, say, prev and candidate.
+     * -    remove any non-overlapping batches from sortedBatches encountered during the find operation and add them to a finalBatchList
+     * -    based on various conditions of offset overlap and batch state differences combine the batches or
+     *      create new batches, if required, and add to the sortedSet.
+     * - done
+     * - return the finalBatchList
+     *
+     * @return list of {@link PersisterStateBatch} representing non-overlapping combined batches
+     */
+    public List<PersisterStateBatch> combineStateBatches() {
+        pruneBatches();
+        mergeBatches();
+        return finalBatchList;
+    }
+
+    private void mergeBatches() {
+        if (combinedBatchList.size() < 2) {
+            finalBatchList = combinedBatchList;
+            return;
+        }
+
+        sortedBatches = new TreeSet<>(combinedBatchList);
```

Comment:
This was done here as the treeset might not be needed at all if intervals are less than 2, hence a small optimization.

Reply:
This code is called from other places as well.
For instance, no assumptions can be made about the incoming request.  They can have overlapping and expired batches. In that case, there is no currentBatches/batchesSoFar as such. We only want to clean up the incoming batches. This is necessary as we will write such requests to the internal topic and the records should be clean.
Since number of snapshot updates is configurable, we can tweak the number to get to a sweet spot with the treeset overhead. Additionally, the share snapshot creation will prune the current batches coupled with expiry on basis of startOffset.